crates/hdk/src/hash.rs

```diff
@@ -1,3 +1,109 @@
+//!
+//! Byte 1 and 2 are taken up by the hash type code as an [unsigned varint](https://github.com/multiformats/unsigned-varint).
+//! Varint is a serial encoding of an integer as a byte array of variable length.
+//! When decoded to a regular integer, varint `132 32` equates to `4100`. This and the other Multihash values employed for Holochain hashes meet several criteria:
+//!
+//! * It encodes as more than one byte, as one byte entries are reserved in Multihash.
+//! * An encoding consisting of two bytes plus the length byte makes three bytes, which always translates to 4 characters in Base64 encoding.
+//! * The resulting Base64 encoding is supposed to be human-recognizable. `hC` was chosen in accordance with `holoChain`.
+//!
+//! Byte 3, which is `0x24` in hexadecimal and `36` in decimal notation, reflects the hash size in bytes, meaning the **digests are 36 bytes long**.
+//!
+//!
+//! ## Valid Holochain hash types
+//!
+//! Here is a list of all valid hash types in Holochain, in hexadecimal, decimal and Base64 notation and what they are used for:
+//!
+//! | hex      | decimal   | base64 | integer |  usage  |
+//! | -------- | --------- | ------ | ------- | ------- |
+//! | 84 20 24 | 132 32 36 | hCAk   | 4100    | Agent   |
+//! | 84 21 24 | 132 33 36 | hCEk   | 4228    | Entry   |
+//! | 84 22 24 | 132 34 36 | hCIk   | 4356    | Net ID  |
+//! | 84 23 24 | 132 35 36 | hCMk   | 4484    |         |
+//! | 84 24 24 | 132 36 36 | hCQk   | 4612    | DHT Op  |
+//! | 84 25 24 | 132 37 36 | hCUk   | 4740    |         |
+//! | 84 26 24 | 132 38 36 | hCYk   | 4868    |         |
+//! | 84 27 24 | 132 39 36 | hCck   | 4996    |         |
+//! | 84 28 24 | 132 40 36 | hCgk   | 5124    |         |
+//! | 84 29 24 | 132 41 36 | hCkk   | 5252    | Header  |
+//! | 84 2a 24 | 132 42 36 | hCok   | 5380    | WASM    |
+//! | 84 2b 24 | 132 43 36 | hCsk   | 5508    |         |
+//! | 84 2c 24 | 132 44 36 | hCwk   | 5636    |         |
+//! | 84 2d 24 | 132 45 36 | hC0k   | 5764    | DNA     |
+//! | 84 2e 24 | 132 46 36 | hC4k   | 5892    |         |
+//! | 84 2f 24 | 132 47 36 | hC8k   | 6020    |         |
+//!
+//!
+//! ### Breakdowns of example
+//!
+//! Breakdown of the example agent pub key as byte array in decimal notation:
+//!
+//! | 132 32           | 36            | 39 218 126 34 87 204 165 227 255 29 236 160 66 221 163 168 112 215 187 143 152 68 4 30 206 173 203 210 111 103 207 124 2 107 67 33        |
+//! | ---------------- | ------------- | ----------------------------------------------------------------------------------------------------------------------------------------- |
+//! | type             | length        | hash                                                                                                                                      |
+//! | public agent key | 36 bytes long | Blake2b hash, 36 bytes long                                                                                                               |
```

Comment:
> Blake2b hash, 36 bytes long

The blake2b hash is only the first 32 bytes. The final 4 are the dht "location" bytes, which can also serve as an integrity check of the hash itself. We interpret these final 4 bytes as a `u32` to identify a particular agent's "neighborhood" in a circle defined by this wrapping u32 value, similarly, the location of an entry or header is used to determine who is responsible for the value by mapping it to the same space.

The location is calculated by taking a 128 bit (16 byte) blake2b hash of the preceding 32 bytes, then progressively xor-ing them into just 4 bytes:

```rust
/// internal compute the holo dht location u32
pub fn holo_dht_location_bytes(data: &[u8]) -> Vec<u8> {
    // Assert the data size is relatively small so we are
    // comfortable executing this synchronously / blocking tokio thread.
    assert_eq!(32, data.len(), "only 32 byte hashes supported");

    let hash = blake2b_128(data);
    let mut out = vec![hash[0], hash[1], hash[2], hash[3]];
    for i in (4..16).step_by(4) {
        out[0] ^= hash[i];
        out[1] ^= hash[i + 1];
        out[2] ^= hash[i + 2];
        out[3] ^= hash[i + 3];
    }
    out
}
```

Reply:
the u32 location defines where on the circle an agent's arc is. currently, it defines the center of the arc, michael is currently working on code that will switch it to defining the left side of the arc

"neighborhood" is a very loose term, roughly meaning a group of agents that have overlapping arcs

"authority" is actually related to dht ops, so yeah, what i actually meant was if you get the location u32 from whatever hash corresponds to the dht op and it happens to fall within an agent's defined arc